File: src/codegraphcontext/tools/languages/java.py
from pathlib import Path
from typing import Any, Dict, Optional, Tuple
import re
from codegraphcontext.utils.debug_log import debug_log, info_logger, error_logger, warning_logger
from codegraphcontext.utils.tree_sitter_manager import execute_query

JAVA_QUERIES = {
    "functions": """
        (method_declaration
            name: (identifier) @name
            parameters: (formal_parameters) @params
        ) @function_node
        
        (constructor_declaration
            name: (identifier) @name
            parameters: (formal_parameters) @params
        ) @function_node
    """,
    "classes": """
        [
            (class_declaration name: (identifier) @name)
            (interface_declaration name: (identifier) @name)
            (enum_declaration name: (identifier) @name)
            (annotation_type_declaration name: (identifier) @name)
        ] @class
    """,
    "imports": """
        (import_declaration) @import
    """,
    "calls": """
        (method_invocation
            name: (identifier) @name
        ) @call_node
        
        (object_creation_expression
            type: [
                (type_identifier)
                (scoped_type_identifier)
                (generic_type)
            ] @name
        ) @call_node
    """,
    "variables": """
        (local_variable_declaration
            type: (_) @type
            declarator: (variable_declarator
                name: (identifier) @name
            )
        ) @variable
        
        (field_declaration
            type: (_) @type
            declarator: (variable_declarator
                name: (identifier) @name
            )
        ) @variable
    """,
}

class JavaTreeSitterParser:
    def __init__(self, generic_parser_wrapper: Any):
        self.generic_parser_wrapper = generic_parser_wrapper
        self.language_name = "java"
        self.language = generic_parser_wrapper.language
        self.parser = generic_parser_wrapper.parser

    def parse(self, path: Path, is_dependency: bool = False, index_source: bool = False) -> Dict[str, Any]:
        try:
            self.index_source = index_source
            with open(path, 'r', encoding='utf-8', errors='ignore') as f:
                source_code = f.read()

            if not source_code.strip():
                warning_logger(f"Empty or whitespace-only file: {path}")
                return {
                    "path": str(path),
                    "functions": [],
                    "classes": [],
                    "variables": [],
                    "imports": [],
                    "function_calls": [],
                    "is_dependency": is_dependency,
                    "lang": self.language_name,
                }

            tree = self.parser.parse(bytes(source_code, "utf8"))

            parsed_functions = []
            parsed_classes = []
            parsed_variables = []
            parsed_imports = []
            parsed_calls = []

            for capture_name, query in JAVA_QUERIES.items():
                results = execute_query(self.language, query, tree.root_node)

                if capture_name == "functions":
                    parsed_functions = self._parse_functions(results, source_code, path)
                elif capture_name == "classes":
                    parsed_classes = self._parse_classes(results, source_code, path)
                elif capture_name == "imports":
                    parsed_imports = self._parse_imports(results, source_code)
                elif capture_name == "calls":
                    parsed_calls = self._parse_calls(results, source_code)
                elif capture_name == "variables":
                    # results for variables query
                    parsed_variables = self._parse_variables(results, source_code, path)

            return {
                "path": str(path),
                "functions": parsed_functions,
                "classes": parsed_classes,
                "variables": parsed_variables,
                "imports": parsed_imports,
                "function_calls": parsed_calls,
                "is_dependency": is_dependency,
                "lang": self.language_name,
            }

        except Exception as e:
            error_logger(f"Error parsing Java file {path}: {e}")
            return {
                "path": str(path),
                "functions": [],
                "classes": [],
                "variables": [],
                "imports": [],
                "function_calls": [],
                "is_dependency": is_dependency,
                "lang": self.language_name,
            }

    def _get_parent_context(self, node: Any) -> Tuple[Optional[str], Optional[str], Optional[int]]:
        curr = node.parent
        while curr:
            if curr.type in ("method_declaration", "constructor_declaration"):
                name_node = curr.child_by_field_name("name")
                return (
                    self._get_node_text(name_node) if name_node else None,
                    curr.type,
                    curr.start_point[0] + 1,
                )
            if curr.type in ("class_declaration", "interface_declaration", "enum_declaration", "annotation_type_declaration"):
                name_node = curr.child_by_field_name("name")
                return (
                    self._get_node_text(name_node) if name_node else None,
                    curr.type,
                    curr.start_point[0] + 1,
                )
            curr = curr.parent
        return None, None, None

    def _get_node_text(self, node: Any) -> str:
        if not node: return ""
        return node.text.decode("utf-8")

    def _parse_functions(self, captures: list, source_code: str, path: Path) -> list[Dict[str, Any]]:
        functions = []
        # Group by node identity or stable key to avoid duplicates
        seen_nodes = set()

        for node, capture_name in captures:
            if capture_name == "function_node":
                node_id = (node.start_byte, node.end_byte, node.type)
                if node_id in seen_nodes:
                    continue
                seen_nodes.add(node_id)
                
                try:
                    start_line = node.start_point[0] + 1
                    end_line = node.end_point[0] + 1
                    
                    name_node = node.child_by_field_name("name")
                    if name_node:
                        func_name = self._get_node_text(name_node)
                        
                        params_node = node.child_by_field_name("parameters")
                        parameters = []
                        if params_node:
                            params_text = self._get_node_text(params_node)
                            parameters = self._extract_parameter_names(params_text)

                        source_text = self._get_node_text(node)
                        
                        # Get class context
                        context_name, context_type, context_line = self._get_parent_context(node)

                        func_data = {
                            "name": func_name,
                            "parameters": parameters,
                            "line_number": start_line,
                            "end_line": end_line,
                            "path": str(path),
                            "lang": self.language_name,
                            "context": context_name,
                            "class_context": context_name if context_type and "class" in context_type else None
                        }

                        if self.index_source:
                            func_data["source"] = source_text
                        
                        functions.append(func_data)
                        
                except Exception as e:
                    error_logger(f"Error parsing function in {path}: {e}")
                    continue

        return functions

    def _parse_classes(self, captures: list, source_code: str, path: Path) -> list[Dict[str, Any]]:
        classes = []
        seen_nodes = set()

        for node, capture_name in captures:
            if capture_name == "class":
                node_id = (node.start_byte, node.end_byte, node.type)
                if node_id in seen_nodes:
                    continue
                seen_nodes.add(node_id)
                
                try:
                    start_line = node.start_point[0] + 1
                    end_line = node.end_point[0] + 1
                    
                    name_node = node.child_by_field_name("name")
                    if name_node:
                        class_name = self._get_node_text(name_node)
                        source_text = self._get_node_text(node)
                        
                        bases = []
                        # Look for superclass (extends)
                        superclass_node = node.child_by_field_name('superclass')
                        if superclass_node:
                            # In Java, superclass field usually points to a type node
                            bases.append(self._get_node_text(superclass_node))

                        # Look for super_interfaces (implements)
                        interfaces_node = node.child_by_field_name('interfaces')
                        if not interfaces_node:
                            interfaces_node = next((c for c in node.children if c.type == 'super_interfaces'), None)
                        
                        if interfaces_node:
                            type_list = interfaces_node.child_by_field_name('list')
                            if not type_list:
                                type_list = next((c for c in interfaces_node.children if c.type == 'type_list'), None)
                            
                            if type_list:
                                for child in type_list.children:
                                    if child.type in ('type_identifier', 'generic_type', 'scoped_type_identifier'):
                                        bases.append(self._get_node_text(child))
                            else:
                                for child in interfaces_node.children:
                                    if child.type in ('type_identifier', 'generic_type', 'scoped_type_identifier'):
                                        bases.append(self._get_node_text(child))

                        class_data = {
                            "name": class_name,
                            "line_number": start_line,
                            "end_line": end_line,
                            "bases": bases,
                            "path": str(path),
                            "lang": self.language_name,
                        }

                        if self.index_source:
                            class_data["source"] = source_text
                        
                        classes.append(class_data)
                        
                except Exception as e:
                    error_logger(f"Error parsing class in {path}: {e}")
                    continue

        return classes

    def _parse_variables(self, captures: list, source_code: str, path: Path) -> list[Dict[str, Any]]:
        variables = []
        seen_vars = set()
        
        for node, capture_name in captures:
            if capture_name == "variable":
                # The capture is on the whole declaration, we look for name/type children or captures
                # But our query captures @name and @type separately on subnodes.
                # Actually, the query structure:
                # (local_variable_declaration ... declarator: (variable_declarator name: (identifier) @name)) @variable
                # This means we get 'variable', 'type', 'name' captures in sequence.
                # We should iterate and group them.
                pass

        # Re-approach: Iterate captures and collect finding.
        # Tree sitter returns a list of (node, capture_name).
        
        # Simpler approach: Iterate 'name' captures that are inside a variable declaration context
        
        current_var = {}
        
        for node, capture_name in captures:
            if capture_name == "name":
                # Check parent to confirm it's a variable declarator
                if node.parent.type == "variable_declarator":
                     var_name = self._get_node_text(node)
                     start_line = node.start_point[0] + 1
                     
                     # Get type? Type is sibling of declarator usually, or child of declaration
                     # local_variable_declaration -> type, variable_declarator
                     declaration = node.parent.parent
                     type_node = declaration.child_by_field_name("type")
                     var_type = self._get_node_text(type_node) if type_node else "Unknown"
                     
                     start_byte = node.start_byte
                     if start_byte in seen_vars:
                         continue
                     seen_vars.add(start_byte)
                     
                     ctx_name, ctx_type, ctx_line = self._get_parent_context(node)

                     variables.append({
                        "name": var_name,
                        "type": var_type,
                        "line_number": start_line,
                        "path": str(path),
                        "lang": self.language_name,
                        "context": ctx_name,
                        "class_context": ctx_name if ctx_type and "class" in ctx_type else None
                     })

        return variables

    def _parse_imports(self, captures: list, source_code: str) -> list[dict]:
        imports = []
        
        for node, capture_name in captures:
            if capture_name == "import":
                try:
                    import_text = self._get_node_text(node)
                    import_match = re.search(r'import\s+(?:static\s+)?([^;]+)', import_text)
                    if import_match:
                        import_path = import_match.group(1).strip()
                        
                        import_data = {
                            "name": import_path,
                            "full_import_name": import_path,
                            "line_number": node.start_point[0] + 1,
                            "alias": None,
                            "context": (None, None),
                            "lang": self.language_name,
                            "is_dependency": False,
                        }
                        imports.append(import_data)
                except Exception as e:
                    error_logger(f"Error parsing import: {e}")
                    continue

        return imports

    def _parse_calls(self, captures: list, source_code: str) -> list[dict]:
        calls = []
        seen_calls = set()
        
        debug_log(f"Processing {len(captures)} captures for calls")

        for node, capture_name in captures:
            if capture_name == "name":
                try:
                    call_name = self._get_node_text(node)
                    line_number = node.start_point[0] + 1
                    
                    # Ensure we identify the full call node
                    call_node = node.parent
                    while call_node and call_node.type not in ("method_invocation", "object_creation_expression"):
                        call_node = call_node.parent
                    
                    if not call_node:
                         # fallback if we matched a loose identifier
                         call_node = node

                    # Avoid duplicates
                    call_key = f"{call_name}_{line_number}"
                    if call_key in seen_calls:
                        continue
                    seen_calls.add(call_key)
                    
                    # Extract arguments
                    args = []
                    if call_node:
                        args_node = next((c for c in call_node.children if c.type == 'argument_list'), None)
                        if args_node:
                            for arg in args_node.children:
                                if arg.type not in ('(', ')', ','):
                                    args.append(self._get_node_text(arg))

                    # Extract meaningful full_name
                    full_name = call_name
                    if call_node.type == 'method_invocation':
                        obj_node = call_node.child_by_field_name('object')
                        if obj_node:
                             full_name = f"{self._get_node_text(obj_node)}.{call_name}"
                    elif call_node.type == 'object_creation_expression':
                        type_node = call_node.child_by_field_name('type')
                        if type_node:
                            full_name = self._get_node_text(type_node)
                    
                    ctx_name, ctx_type, ctx_line = self._get_parent_context(node)
                    
                    debug_log(f"Found call: {call_name} (full_name: {full_name}, args: {args}) in context {ctx_name}")

                    call_data = {
                        "name": call_name,
                        "full_name": full_name,
                        "line_number": line_number,
                        "args": args,
                        "inferred_obj_type": None,
                        "context": (ctx_name, ctx_type, ctx_line),
                        "class_context": (ctx_name, ctx_line) if ctx_type and "class" in ctx_type else (None, None),
                        "lang": self.language_name,
                        "is_dependency": False,
                    }
                    calls.append(call_data)
                except Exception as e:
                    error_logger(f"Error parsing call: {e}")
                    continue

        return calls
    

    def _extract_parameter_names(self, params_text: str) -> list[str]:
        params = []
        if not params_text or params_text.strip() == "()":
            return params
            
        params_content = params_text.strip("()")
        if not params_content:
            return params
            
        for param in params_content.split(","):
            param = param.strip()
            if param:
                parts = param.split()
                if len(parts) >= 2:
                    param_name = parts[-1]
                    params.append(param_name)
                    
        return params


def pre_scan_java(files: list[Path], parser_wrapper) -> dict:
    name_to_files = {}
    
    for path in files:
        try:
            with open(path, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
            
            class_matches = re.finditer(r'\b(?:public\s+|private\s+|protected\s+)?(?:static\s+)?(?:abstract\s+)?(?:final\s+)?class\s+(\w+)', content)
            for match in class_matches:
                class_name = match.group(1)
                if class_name not in name_to_files:
                    name_to_files[class_name] = []
                name_to_files[class_name].append(str(path))
            
            interface_matches = re.finditer(r'\b(?:public\s+|private\s+|protected\s+)?interface\s+(\w+)', content)
            for match in interface_matches:
                interface_name = match.group(1)
                if interface_name not in name_to_files:
                    name_to_files[interface_name] = []
                name_to_files[interface_name].append(str(path))
                
        except Exception as e:
            error_logger(f"Error pre-scanning Java file {path}: {e}")
            
    return name_to_files

File: rust/cgc-core/src/lang/java.rs
use std::collections::HashSet;

use streaming_iterator::StreamingIterator;
use tree_sitter::{Language, Node, Query, QueryCursor};

use super::{get_node_text, get_parent_context, LanguageExtractor};
use crate::types::*;

const COMPLEXITY_TYPES: &[&str] = &[
    "if_statement",
    "for_statement",
    "while_statement",
    "do_statement",
    "switch_expression",
    "catch_clause",
    "conditional_expression",
];

const QUERY_FUNCTIONS: &str = r#"
    (method_declaration
        name: (identifier) @name
        parameters: (formal_parameters) @params
    ) @function_node

    (constructor_declaration
        name: (identifier) @name
        parameters: (formal_parameters) @params
    ) @function_node
"#;

const QUERY_CLASSES: &str = r#"
    [
        (class_declaration name: (identifier) @name)
        (interface_declaration name: (identifier) @name)
        (enum_declaration name: (identifier) @name)
        (annotation_type_declaration name: (identifier) @name)
    ] @class
"#;

const QUERY_IMPORTS: &str = r#"
    (import_declaration) @import
"#;

const QUERY_CALLS: &str = r#"
    (method_invocation
        name: (identifier) @name
    ) @call_node

    (object_creation_expression
        type: [
            (type_identifier)
            (scoped_type_identifier)
            (generic_type)
        ] @name
    ) @call_node
"#;

const QUERY_VARIABLES: &str = r#"
    (local_variable_declaration
        type: (_) @type
        declarator: (variable_declarator
            name: (identifier) @name
        )
    ) @variable

    (field_declaration
        type: (_) @type
        declarator: (variable_declarator
            name: (identifier) @name
        )
    ) @variable
"#;

const QUERY_PRE_SCAN: &str = r#"
    (class_declaration name: (identifier) @name)
    (interface_declaration name: (identifier) @name)
"#;

pub struct JavaExtractor;

impl JavaExtractor {
    fn execute_query<'a>(
        &self,
        query_str: &str,
        root: &'a Node<'a>,
        source: &'a [u8],
    ) -> Vec<(Node<'a>, String)> {
        let lang = self.language();
        let query = match Query::new(&lang, query_str) {
            Ok(q) => q,
            Err(_) => return Vec::new(),
        };
        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&query, *root, source);
        let capture_names = query.capture_names();

        let mut results = Vec::new();
        while let Some(m) = matches.next() {
            for cap in m.captures {
                let name = &capture_names[cap.index as usize];
                results.push((cap.node, name.to_string()));
            }
        }
        results
    }

    fn get_parent_context_java(
        &self,
        node: &Node,
        source: &[u8],
    ) -> (Option<String>, Option<String>, Option<usize>) {
        let types = &[
            "method_declaration",
            "constructor_declaration",
            "class_declaration",
            "interface_declaration",
            "enum_declaration",
            "annotation_type_declaration",
        ];
        get_parent_context(node, source, types)
    }

    fn get_class_context(
        &self,
        node: &Node,
        source: &[u8],
    ) -> (Option<String>, Option<String>) {
        let types = &[
            "class_declaration",
            "interface_declaration",
            "enum_declaration",
            "annotation_type_declaration",
        ];
        let (name, kind, _) = get_parent_context(node, source, types);
        (name, kind)
    }

    /// Extract parameter names from a formal_parameters text like "(int x, String name)"
    fn extract_parameter_names(&self, params_text: &str) -> Vec<String> {
        let mut params = Vec::new();
        let content = params_text.trim().trim_start_matches('(').trim_end_matches(')');
        if content.is_empty() {
            return params;
        }
        for param in content.split(',') {
            let param = param.trim();
            if param.is_empty() {
                continue;
            }
            let parts: Vec<&str> = param.split_whitespace().collect();
            if parts.len() >= 2 {
                params.push(parts.last().unwrap().to_string());
            }
        }
        params
    }
}

impl LanguageExtractor for JavaExtractor {
    fn language(&self) -> Language {
        tree_sitter_java::LANGUAGE.into()
    }

    fn lang_name(&self) -> &str {
        "java"
    }

    fn complexity_node_types(&self) -> &[&str] {
        COMPLEXITY_TYPES
    }

    fn find_functions(
        &self,
        root: &Node,
        source: &[u8],
        index_source: bool,
    ) -> Vec<FunctionData> {
        let mut functions = Vec::new();
        let mut seen_nodes: HashSet<(usize, usize)> = HashSet::new();

        for (node, capture_name) in self.execute_query(QUERY_FUNCTIONS, root, source) {
            if capture_name != "function_node" {
                continue;
            }

            let node_id = (node.start_byte(), node.end_byte());
            if seen_nodes.contains(&node_id) {
                continue;
            }
            seen_nodes.insert(node_id);

            let name_node = match node.child_by_field_name("name") {
                Some(n) => n,
                None => continue,
            };
            let func_name = get_node_text(&name_node, source).to_string();

            let params = node
                .child_by_field_name("parameters")
                .map(|pn| {
                    let params_text = get_node_text(&pn, source);
                    self.extract_parameter_names(params_text)
                })
                .unwrap_or_default();

            let (context, context_type, _) = self.get_parent_context_java(&node, source);
            let (class_name, class_type) = self.get_class_context(&node, source);
            let class_context = if class_type.is_some() {
                class_name
            } else {
                None
            };

            let complexity = self.calculate_complexity(&node);

            let mut func = FunctionData {
                name: func_name,
                line_number: node.start_position().row + 1,
                end_line: node.end_position().row + 1,
                args: params,
                cyclomatic_complexity: complexity,
                context,
                context_type,
                class_context,
                decorators: Vec::new(),
                lang: self.lang_name().to_string(),
                is_dependency: false,
                source: None,
                docstring: None,
            };

            if index_source {
                func.source = Some(get_node_text(&node, source).to_string());
            }

            functions.push(func);
        }

        functions
    }

    fn find_classes(
        &self,
        root: &Node,
        source: &[u8],
        index_source: bool,
    ) -> Vec<ClassData> {
        let mut classes = Vec::new();
        let mut seen_nodes: HashSet<(usize, usize)> = HashSet::new();

        for (node, capture_name) in self.execute_query(QUERY_CLASSES, root, source) {
            if capture_name != "class" {
                continue;
            }

            let node_id = (node.start_byte(), node.end_byte());
            if seen_nodes.contains(&node_id) {
                continue;
            }
            seen_nodes.insert(node_id);

            let name_node = match node.child_by_field_name("name") {
                Some(n) => n,
                None => continue,
            };
            let class_name = get_node_text(&name_node, source).to_string();

            let mut bases = Vec::new();

            // Look for superclass (extends)
            if let Some(superclass_node) = node.child_by_field_name("superclass") {
                bases.push(get_node_text(&superclass_node, source).to_string());
            }

            // Look for interfaces (implements) - try field name first, then scan children
            let interfaces_node = node
                .child_by_field_name("interfaces")
                .or_else(|| {
                    (0..node.child_count())
                        .filter_map(|i| node.child(i))
                        .find(|c| c.kind() == "super_interfaces")
                });

            if let Some(ifaces) = interfaces_node {
                // Find type_list within
                let type_list = ifaces
                    .child_by_field_name("list")
                    .or_else(|| {
                        (0..ifaces.child_count())
                            .filter_map(|i| ifaces.child(i))
                            .find(|c| c.kind() == "type_list")
                    });

                if let Some(tl) = type_list {
                    for i in 0..tl.child_count() {
                        if let Some(child) = tl.child(i) {
                            if matches!(
                                child.kind(),
                                "type_identifier" | "generic_type" | "scoped_type_identifier"
                            ) {
                                bases.push(get_node_text(&child, source).to_string());
                            }
                        }
                    }
                } else {
                    // Fallback: scan children of interfaces node directly
                    for i in 0..ifaces.child_count() {
                        if let Some(child) = ifaces.child(i) {
                            if matches!(
                                child.kind(),
                                "type_identifier" | "generic_type" | "scoped_type_identifier"
                            ) {
                                bases.push(get_node_text(&child, source).to_string());
                            }
                        }
                    }
                }
            }

            let (context, _, _) = get_parent_context(
                &node,
                source,
                &["class_declaration", "interface_declaration"],
            );

            let mut class = ClassData {
                name: class_name,
                line_number: node.start_position().row + 1,
                end_line: node.end_position().row + 1,
                bases,
                context,
                decorators: Vec::new(),
                lang: self.lang_name().to_string(),
                is_dependency: false,
                source: None,
                docstring: None,
            };

            if index_source {
                class.source = Some(get_node_text(&node, source).to_string());
            }

            classes.push(class);
        }

        classes
    }

    fn find_imports(&self, root: &Node, source: &[u8]) -> Vec<ImportData> {
        let mut imports = Vec::new();

        for (node, capture_name) in self.execute_query(QUERY_IMPORTS, root, source) {
            if capture_name != "import" {
                continue;
            }

            let import_text = get_node_text(&node, source);

            // Parse: import [static] path;
            let trimmed = import_text.trim();
            let path = if let Some(rest) = trimmed.strip_prefix("import") {
                let rest = rest.trim();
                let rest = if let Some(r) = rest.strip_prefix("static") {
                    r.trim()
                } else {
                    rest
                };
                rest.trim_end_matches(';').trim().to_string()
            } else {
                continue;
            };

            if path.is_empty() {
                continue;
            }

            imports.push(ImportData {
                name: path.clone(),
                full_import_name: path,
                line_number: node.start_position().row + 1,
                alias: None,
                context: (None, None),
                lang: self.lang_name().to_string(),
                is_dependency: false,
            });
        }

        imports
    }

    fn find_calls(&self, root: &Node, source: &[u8]) -> Vec<CallData> {
        let mut calls = Vec::new();
        let mut seen_calls: HashSet<String> = HashSet::new();

        for (node, capture_name) in self.execute_query(QUERY_CALLS, root, source) {
            if capture_name != "name" {
                continue;
            }

            let call_name = get_node_text(&node, source).to_string();
            let line_number = node.start_position().row + 1;

            let call_key = format!("{}_{}", call_name, line_number);
            if seen_calls.contains(&call_key) {
                continue;
            }
            seen_calls.insert(call_key);

            // Navigate up to call_node
            let mut call_node = node.parent();
            while let Some(cn) = call_node {
                if cn.kind() == "method_invocation"
                    || cn.kind() == "object_creation_expression"
                {
                    break;
                }
                call_node = cn.parent();
            }
            let call_node = match call_node {
                Some(cn) => cn,
                None => continue,
            };

            // Extract args
            let mut args = Vec::new();
            let args_node = (0..call_node.child_count())
                .filter_map(|i| call_node.child(i))
                .find(|c| c.kind() == "argument_list");
            if let Some(an) = args_node {
                for i in 0..an.child_count() {
                    if let Some(arg) = an.child(i) {
                        let text = get_node_text(&arg, source);
                        if !text.is_empty()
                            && text != "("
                            && text != ")"
                            && text != ","
                        {
                            args.push(text.to_string());
                        }
                    }
                }
            }

            // Determine full_name and inferred_obj_type
            let mut full_name = call_name.clone();
            let mut inferred_obj_type = None;

            if call_node.kind() == "method_invocation" {
                if let Some(obj_node) = call_node.child_by_field_name("object") {
                    let obj_text = get_node_text(&obj_node, source);
                    full_name = format!("{}.{}", obj_text, call_name);
                    // Simple identifier resolution
                    let base_obj = obj_text.split('.').next().unwrap_or(obj_text);
                    if !base_obj.contains('(') {
                        inferred_obj_type = Some(base_obj.to_string());
                    }
                }
            } else if call_node.kind() == "object_creation_expression" {
                if let Some(type_node) = call_node.child_by_field_name("type") {
                    full_name = get_node_text(&type_node, source).to_string();
                }
            }

            let context = self.get_parent_context_java(&node, source);
            let (class_name, class_type) = self.get_class_context(&node, source);

            calls.push(CallData {
                name: call_name,
                full_name,
                line_number,
                args,
                inferred_obj_type,
                context,
                class_context: (class_name, class_type),
                lang: self.lang_name().to_string(),
                is_dependency: false,
                is_indirect_call: false,
            });
        }

        calls
    }

    fn find_variables(&self, root: &Node, source: &[u8]) -> Vec<VariableData> {
        let mut variables = Vec::new();
        let mut seen_vars: HashSet<usize> = HashSet::new();

        for (node, capture_name) in self.execute_query(QUERY_VARIABLES, root, source) {
            if capture_name != "name" {
                continue;
            }

            let start_byte = node.start_byte();
            if seen_vars.contains(&start_byte) {
                continue;
            }
            seen_vars.insert(start_byte);

            // node is the identifier inside variable_declarator
            // Walk up: variable_declarator -> declaration
            let declaration = match node.parent().and_then(|p| p.parent()) {
                Some(d) => d,
                None => continue,
            };

            let type_annotation = declaration
                .child_by_field_name("type")
                .map(|tn| get_node_text(&tn, source).to_string());

            let var_name = get_node_text(&node, source).to_string();

            let (context, context_type, _) = self.get_parent_context_java(&node, source);
            let (class_name, class_type) = self.get_class_context(&node, source);
            let class_context = if class_type.is_some() {
                class_name
            } else {
                None
            };

            variables.push(VariableData {
                name: var_name,
                line_number: node.start_position().row + 1,
                value: None,
                type_annotation,
                context,
                class_context,
                lang: self.lang_name().to_string(),
                is_dependency: false,
            });
        }

        variables
    }

    fn pre_scan_definitions(&self, root: &Node, source: &[u8]) -> Vec<String> {
        let mut names = Vec::new();
        for (node, _) in self.execute_query(QUERY_PRE_SCAN, root, source) {
            names.push(get_node_text(&node, source).to_string());
        }
        names
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tree_sitter::Parser;

    fn parse_source(code: &str) -> (tree_sitter::Tree, Vec<u8>) {
        let mut parser = Parser::new();
        let lang: Language = tree_sitter_java::LANGUAGE.into();
        parser.set_language(&lang).unwrap();
        let source = code.as_bytes().to_vec();
        let tree = parser.parse(&source, None).unwrap();
        (tree, source)
    }

    #[test]
    fn test_find_functions() {
        let code = r#"
public class MyClass {
    public void hello(String name, int age) {
        System.out.println(name);
    }

    public int add(int a, int b) {
        return a + b;
    }
}
"#;
        let (tree, source) = parse_source(code);
        let ext = JavaExtractor;
        let funcs = ext.find_functions(&tree.root_node(), &source, false);
        assert_eq!(funcs.len(), 2);

        let hello = funcs.iter().find(|f| f.name == "hello").unwrap();
        assert_eq!(hello.args, vec!["name", "age"]);

        let add = funcs.iter().find(|f| f.name == "add").unwrap();
        assert_eq!(add.args, vec!["a", "b"]);
    }

    #[test]
    fn test_find_classes() {
        let code = r#"
public class Animal {
    String name;
}

public class Dog extends Animal {
    public void bark() {}
}
"#;
        let (tree, source) = parse_source(code);
        let ext = JavaExtractor;
        let classes = ext.find_classes(&tree.root_node(), &source, false);
        assert_eq!(classes.len(), 2);
        assert_eq!(classes[0].name, "Animal");
        assert!(classes[0].bases.is_empty());
        assert_eq!(classes[1].name, "Dog");
        assert!(!classes[1].bases.is_empty());
    }

    #[test]
    fn test_find_imports() {
        let code = r#"
import java.util.List;
import static java.lang.Math.PI;
import java.io.*;
"#;
        let (tree, source) = parse_source(code);
        let ext = JavaExtractor;
        let imports = ext.find_imports(&tree.root_node(), &source);
        assert!(imports.len() >= 2);
        assert_eq!(imports[0].name, "java.util.List");
        assert_eq!(imports[1].name, "java.lang.Math.PI");
    }

    #[test]
    fn test_find_calls() {
        let code = r#"
public class Main {
    public void run() {
        System.out.println("hello");
        List<String> items = new ArrayList<>();
    }
}
"#;
        let (tree, source) = parse_source(code);
        let ext = JavaExtractor;
        let calls = ext.find_calls(&tree.root_node(), &source);
        assert!(calls.len() >= 1);
        assert!(calls.iter().any(|c| c.name == "println"));
    }

    #[test]
    fn test_find_variables() {
        let code = r#"
public class Main {
    private int count;
    public void run() {
        String name = "hello";
    }
}
"#;
        let (tree, source) = parse_source(code);
        let ext = JavaExtractor;
        let vars = ext.find_variables(&tree.root_node(), &source);
        assert!(vars.len() >= 2);
        assert!(vars.iter().any(|v| v.name == "count"));
        assert!(vars.iter().any(|v| v.name == "name"));
    }

    #[test]
    fn test_pre_scan() {
        let code = r#"
public class MyService {
}

public interface MyInterface {
}
"#;
        let (tree, source) = parse_source(code);
        let ext = JavaExtractor;
        let names = ext.pre_scan_definitions(&tree.root_node(), &source);
        assert!(names.contains(&"MyService".to_string()));
        assert!(names.contains(&"MyInterface".to_string()));
    }
}
